core/arch/src/main/java/com/appcoins/wallet/core/arch/NewBaseViewModel.kt

```diff
@@ -0,0 +1,206 @@
+    } catch (e: Exception) {
+      setState { reducer(Async.Fail(Error.UnknownError(e), retainValue?.get(this)?.value)) }
+    }
+  }
+
+  /**
+   * Maps DataResult<T> to Async<T> and updates the state through [reducer]
+   */
+  private fun <T> DataResult<T>.toAsyncWithState(
+    retainValue: KProperty1<S, Async<T>>? = null, reducer: S.(Async<T>) -> S
+  ) {
+    when (this) {
+      is Ok -> setState { reducer(Async.Success(this@toAsyncWithState.value)) }
+      is Err -> setState {
+        reducer(Async.Fail(error, retainValue?.get(this)?.value))
+      }
+    }
+  }
+
+  /**
+   * Executes this suspend function and wraps the its state in Async. It automatically emits
+   * Async.Loading on execute and Async.Fail on uncaught exception.
+   *
+   * Example usage:
+   *   viewModelScope.launch {
+   *     suspend { interactor.getData() }
+   *         .mapSuspendToAsync(HomeState::dataAsync) { async -> copy(dataAsync = async) }
+   *   }
+   *
+   * @param retainValue A state property that will be retained in case of Loading or Fail
+   * @param reducer A reducer that receives the result of this suspend function as well as the
+   *                current state and returns the new state to be set.
+   */
+  protected suspend fun <T : Any?> (suspend () -> T).mapSuspendToAsync(
+    retainValue: KProperty1<S, Async<T>>? = null,
+    reducer: S.(Async<T>) -> S
+  ) {
+    setState { reducer(Async.Loading(retainValue?.get(this)?.value)) }
+    try {
+      val result = this.invoke()
+      setState { reducer(Async.Success(result)) }
+    } catch (e: Exception) {
+      setState { reducer(Async.Fail(Error.UnknownError(e), retainValue?.get(this)?.value)) }
+    }
+  }
+
+  protected fun setState(reducer: S.() -> S) {
+    viewModelScope.launch {
+      _stateFlow.update { state -> state.reducer() }
+    }
+  }
+
+  private fun Throwable.mapToError(): Error {
+    if (this.isNoNetworkException()) {
+      return Error.ApiError.NetworkError(this)
+    }
+    return Error.UnknownError(this)
+  }
+
+  protected fun cancelSubscription(id: String) {
+    repeatableJobsMap[id]?.cancel()
+  }
+
+  @CallSuper
+  override fun onCleared() {
+    repeatableJobsMap.clear()
+  }
+
+  protected fun withState(action: S.() -> Unit) = stateFlow.value.action()
+
+  protected inline fun <A> subscribeState(
+    prop: KProperty1<S, A>,
+    crossinline actionBlock: suspend (A) -> Unit
+  ) {
+    stateFlow
+      .map { s -> prop.get(s) }
+      .distinctUntilChanged()
+      .onEach { a -> actionBlock(a) }
+      .launchIn(viewModelScope)
+  }
+
+  protected inline fun subscribeState(crossinline actionBlock: suspend (S) -> Unit) {
+    stateFlow
+      .onEach { s -> actionBlock(s) }
+      .launchIn(viewModelScope)
+  }
+
+  /**
+   * Launches a job for an [id]. If a job is already running with that [id], it cancels it first
+   * before starting the new job.
+   */
+  protected fun <T> Flow<T>.repeatableLaunchIn(scope: CoroutineScope, id: String): Job {
```

Comment:
why keep two BaseViewModel?

Reply:
The second one (NewBaseViewModel) is for coroutines with result, leaving the current BaseViewModel for rxJava with no Result.
This is mostly to separate the two implementations and deprecate one after a migration